goa gen makes error with CollectionOf and Body

## Expected Behavior

`goa gen` generates an array instead of struct for `ListResponseBody` with DSL like https://github.com/akm/goa_v2_simple_example/blob/master/design/person.go#L36-L57

## Actual Behavior

```
$ goa gen github.com/akm/goa_v2_simple_example/design
exit status 2
res.People is an array but body type is
body is an object but v type is PersonCollection
panic: failed to project media type "application/vnd.person+json; type=collection": unknown view "default"

goroutine 1 [running]:
goa.design/goa/http/codegen/openapi.ResultTypeRefWithPrefix(0xc420162000, 0xc42026a180, 0x165618f, 0x7, 0x1651a64, 0x6, 0x0, 0xc420173278)
	$GOPATH/src/goa.design/goa/http/codegen/openapi/json_schema.go:230 +0x358
goa.design/goa/http/codegen/openapi.responseSpecFromExpr(0xc4203f2000, 0x19dafa0, 0xc4202624d0, 0x1651a64, 0x6, 0x0)
	$GOPATH/src/goa.design/goa/http/codegen/openapi/openapi_v2_builder.go:428 +0x4b2
goa.design/goa/http/codegen/openapi.buildPathFromExpr(0xc4203f2000, 0x19dafa0, 0xc42019bdb0, 0xc420266570, 0x0, 0x0)
	$GOPATH/src/goa.design/goa/http/codegen/openapi/openapi_v2_builder.go:555 +0x285
goa.design/goa/http/codegen/openapi.NewV2(0x19dafa0, 0xc42019bdb0, 0x3, 0xc4200170e0, 0x15)
	$GOPATH/src/goa.design/goa/http/codegen/openapi/openapi_v2_builder.go:91 +0xc0c
goa.design/goa/http/codegen.OpenAPIFiles(0x19dafa0, 0xe, 0x20, 0xc420157680, 0xc42024c850, 0x0)
	$GOPATH/src/goa.design/goa/http/codegen/openapi.go:27 +0x141
goa.design/goa/codegen/generator.OpenAPI(0xc42023c254, 0x28, 0xc42029c140, 0x2, 0x2, 0xa, 0x10, 0xa, 0x0, 0x0)
	$GOPATH/src/goa.design/goa/codegen/generator/openapi.go:16 +0x59
goa.design/goa/codegen/generator.Generate(0x7ffeefbff747, 0x1, 0x1648f67, 0x3, 0x1, 0x0, 0xc42018d300, 0x0, 0xffffffffffffffff)
	$GOPATH/src/goa.design/goa/codegen/generator/generate.go:62 +0x37c
main.main()
	$GOPATH/src/github.com/akm/goa_v2_simple_example/goa364214395/main.go:54 +0x2dd
```

## Steps to Reproduce the Problem (including complete and simple design if relevant)

```
$ mkdir -p $GOPATH/src/github.com/akm
$ cd $GOPATH/src/github.com/akm
$ git clone https://github.com/akm/goa_v2_simple_example.git
$ cd goa_v2_simple_example
$ goa gen github.com/akm/goa_v2_simple_example/design
```

## Note

If `Body` is removed, `goa gen` finishes without any error. 
https://github.com/akm/goa_v2_simple_example/blob/if_remove_body/design/person.go#L54

But `ListResponseBody` generated is not an array but a struct
https://github.com/akm/goa_v2_simple_example/commit/5eabc221188f7570504fa89577f333632bafb1fc#diff-c255176bae3232fc27e5dd12fbb14929


## Specifications

  - goa Version: https://github.com/goadesign/goa/commit/5cadea1d967fe93c7b9ef8bede7ed300b7e40e45
  - Platform (Linux/Windows etc.): Mac OS X 10.14.4 (18E226)

--- a/codegen/cli/cli.go
+++ b/codegen/cli/cli.go
@@ -466,6 +466,7 @@ func conversionCode(from, to, typeName string, pointer bool) (string, bool) {
 		checkErr = true
 	case int64N:
 		parse = fmt.Sprintf("%s, err %s= strconv.ParseInt(%s, 10, 64)", target, decl, from)
+		checkErr = true
 	case uintN:
 		parse = fmt.Sprintf("var v uint64\nv, err = strconv.ParseUint(%s, 10, 64)", from)
 		cast = fmt.Sprintf("%s %s= uint(v)", target, decl)

--- a/dsl/http.go
+++ b/dsl/http.go
@@ -664,6 +664,11 @@ func Body(args ...interface{}) {
 		} else {
 			attr.Meta["origin:attribute"] = []string{a}
 		}
+		if rt, ok := attr.Type.(*expr.ResultTypeExpr); ok {
+			// If the attribute type is a result type add the type to the
+			// GeneratedTypes so that the type's DSLFunc is executed.
+			*expr.Root.GeneratedTypes = append(*expr.Root.GeneratedTypes, rt)
+		}
 	case expr.UserType:
 		attr = &expr.AttributeExpr{Type: a}
 		if len(args) > 1 {

--- a/expr/attribute.go
+++ b/expr/attribute.go
@@ -220,7 +220,7 @@ func (a *AttributeExpr) Validate(ctx string, parent eval.Expression) *eval.Valid
 // attribute.
 func (a *AttributeExpr) Finalize() {
 	if ut, ok := a.Type.(UserType); ok {
-		ut.Attribute().Finalize()
+		ut.Finalize()
 	}
 	if IsObject(a.Type) {
 		for _, ref := range a.References {

--- a/expr/http_endpoint.go
+++ b/expr/http_endpoint.go
@@ -537,6 +537,7 @@ func (e *HTTPEndpointExpr) Finalize() {
 		if r.Body == nil {
 			r.Body = httpResponseBody(e, r)
 		}
+		r.Body.Finalize()
 	}
 
 	// Lookup undefined HTTP errors in API.

--- a/http/codegen/client_decode_test.go
+++ b/http/codegen/client_decode_test.go
@@ -18,6 +18,7 @@ func TestClientDecode(t *testing.T) {
 		{"body-result-multiple-views", testdata.ResultBodyMultipleViewsDSL, testdata.ResultBodyMultipleViewsDecodeCode},
 		{"empty-body-result-multiple-views", testdata.EmptyBodyResultMultipleViewsDSL, testdata.EmptyBodyResultMultipleViewsDecodeCode},
 		{"explicit-body-result-multiple-views", testdata.ExplicitBodyUserResultMultipleViewsDSL, testdata.ExplicitBodyUserResultMultipleViewsDecodeCode},
+		{"explicit-body-result-collection", testdata.ExplicitBodyResultCollectionDSL, testdata.ExplicitBodyResultCollectionDecodeCode},
 		{"tag-result-multiple-views", testdata.ResultMultipleViewsTagDSL, testdata.ResultMultipleViewsTagDecodeCode},
 		{"empty-server-response-with-tags", testdata.EmptyServerResponseWithTagsDSL, testdata.EmptyServerResponseWithTagsDecodeCode},
 	}

--- a/http/codegen/server_encode_test.go
+++ b/http/codegen/server_encode_test.go
@@ -68,6 +68,7 @@ func TestEncode(t *testing.T) {
 
 		{"explicit-body-primitive-result-multiple-views", testdata.ExplicitBodyPrimitiveResultMultipleViewsDSL, testdata.ExplicitBodyPrimitiveResultMultipleViewsEncodeCode},
 		{"explicit-body-user-result-multiple-views", testdata.ExplicitBodyUserResultMultipleViewsDSL, testdata.ExplicitBodyUserResultMultipleViewsEncodeCode},
+		{"explicit-body-result-collection", testdata.ExplicitBodyResultCollectionDSL, testdata.ExplicitBodyResultCollectionEncodeCode},
 		{"explicit-content-type-result", testdata.ExplicitContentTypeResultDSL, testdata.ExplicitContentTypeResultEncodeCode},
 		{"explicit-content-type-response", testdata.ExplicitContentTypeResponseDSL, testdata.ExplicitContentTypeResponseEncodeCode},
 

--- a/http/codegen/service_data.go
+++ b/http/codegen/service_data.go
@@ -2045,7 +2045,12 @@ func buildResponseBodyType(bodyCA, attCA *codegen.ContextualAttribute, e *expr.H
 				// generate validation code for unmarshaled type (client-side).
 				validateDef = codegen.RecursiveValidationCode(bodyCA, "body")
 				if validateDef != "" {
-					validateRef = fmt.Sprintf("err = Validate%s(&body)", varname)
+					target := "&body"
+					if expr.IsArray(ut) {
+						// result type collection
+						target = "body"
+					}
+					validateRef = fmt.Sprintf("err = Validate%s(%s)", varname, target)
 				}
 			}
 		} else if !expr.IsPrimitive(body.Type) && mustInit {

--- a/http/codegen/testdata/result_decode_functions.go
+++ b/http/codegen/testdata/result_decode_functions.go
@@ -172,6 +172,48 @@ func DecodeMethodExplicitBodyUserResultMultipleViewResponse(decoder func(*http.R
 }
 `
 
+var ExplicitBodyResultCollectionDecodeCode = `// DecodeMethodExplicitBodyResultCollectionResponse returns a decoder for
+// responses returned by the ServiceExplicitBodyResultCollection
+// MethodExplicitBodyResultCollection endpoint. restoreBody controls whether
+// the response body should be restored after having been read.
+func DecodeMethodExplicitBodyResultCollectionResponse(decoder func(*http.Response) goahttp.Decoder, restoreBody bool) func(*http.Response) (interface{}, error) {
+	return func(resp *http.Response) (interface{}, error) {
+		if restoreBody {
+			b, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				return nil, err
+			}
+			resp.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+			defer func() {
+				resp.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+			}()
+		} else {
+			defer resp.Body.Close()
+		}
+		switch resp.StatusCode {
+		case http.StatusOK:
+			var (
+				body ResulttypeCollection
+				err  error
+			)
+			err = decoder(resp).Decode(&body)
+			if err != nil {
+				return nil, goahttp.ErrDecodingError("ServiceExplicitBodyResultCollection", "MethodExplicitBodyResultCollection", err)
+			}
+			err = ValidateResulttypeCollection(body)
+			if err != nil {
+				return nil, goahttp.ErrValidationError("ServiceExplicitBodyResultCollection", "MethodExplicitBodyResultCollection", err)
+			}
+			res := NewMethodExplicitBodyResultCollectionResultOK(body)
+			return res, nil
+		default:
+			body, _ := ioutil.ReadAll(resp.Body)
+			return nil, goahttp.ErrInvalidResponse("ServiceExplicitBodyResultCollection", "MethodExplicitBodyResultCollection", resp.StatusCode, string(body))
+		}
+	}
+}
+`
+
 var ResultMultipleViewsTagDecodeCode = `// DecodeMethodTagMultipleViewsResponse returns a decoder for responses
 // returned by the ServiceTagMultipleViews MethodTagMultipleViews endpoint.
 // restoreBody controls whether the response body should be restored after

--- a/http/codegen/testdata/result_dsls.go
+++ b/http/codegen/testdata/result_dsls.go
@@ -771,6 +771,29 @@ var ExplicitBodyUserResultMultipleViewsDSL = func() {
 	})
 }
 
+var ExplicitBodyResultCollectionDSL = func() {
+	var ResultType = ResultType("ResultType", func() {
+		Attributes(func() {
+			Attribute("x", String, func() {
+				MinLength(5)
+			})
+		})
+	})
+	Service("ServiceExplicitBodyResultCollection", func() {
+		Method("MethodExplicitBodyResultCollection", func() {
+			Result(func() {
+				Attribute("a", CollectionOf(ResultType))
+			})
+			HTTP(func() {
+				POST("/")
+				Response(StatusOK, func() {
+					Body("a")
+				})
+			})
+		})
+	})
+}
+
 var ExplicitContentTypeResultDSL = func() {
 	var ResultType = ResultType("ResultType", func() {
 		ContentType("application/custom+json")

--- a/http/codegen/testdata/result_encode_functions.go
+++ b/http/codegen/testdata/result_encode_functions.go
@@ -753,6 +753,20 @@ func EncodeMethodExplicitBodyUserResultMultipleViewResponse(encoder func(context
 }
 `
 
+var ExplicitBodyResultCollectionEncodeCode = `// EncodeMethodExplicitBodyResultCollectionResponse returns an encoder for
+// responses returned by the ServiceExplicitBodyResultCollection
+// MethodExplicitBodyResultCollection endpoint.
+func EncodeMethodExplicitBodyResultCollectionResponse(encoder func(context.Context, http.ResponseWriter) goahttp.Encoder) func(context.Context, http.ResponseWriter, interface{}) error {
+	return func(ctx context.Context, w http.ResponseWriter, v interface{}) error {
+		res := v.(*serviceexplicitbodyresultcollection.MethodExplicitBodyResultCollectionResult)
+		enc := encoder(ctx, w)
+		body := NewResulttypeCollection(res)
+		w.WriteHeader(http.StatusOK)
+		return enc.Encode(body)
+	}
+}
+`
+
 var ExplicitContentTypeResultEncodeCode = `// EncodeMethodExplicitContentTypeResultResponse returns an encoder for
 // responses returned by the ServiceExplicitContentTypeResult
 // MethodExplicitContentTypeResult endpoint.

